### ISSUE TITLE
meson claims Unknown compiler(s): ['c++', 'g++', 'clang++', 'pgc++', 'icpc']

### ISSUE THREAD

**Issue by DimStar77** [mesonbuild/meson#6559]
This is kinda weird: I have a very small dummy meson.build which errors out with:

> meson.build:7:0: ERROR: Unknown compiler(s): ['c++', 'g++', 'clang++', 'pgc++', 'icpc']

Whereas it's correct I don't have any of those compilers on my system, there is also no indication in my meson.build that this would be needed.

The sample meson.build is:

```
project('sample',
	'c',
	version: '1.0.0',
	meson_version: '>= 0.47',
)

dep = dependency('doesnotexist', required: false)
```

**Comment by lb90** [comment]
Hi! Could you please give us more insights on that issue? Specifically, what's your system and which Meson version are you running? Thanks!

**Comment by DimStar77** [comment]
openSUSE Tumbleweed, meson 0.53.1

**Comment by jon-turney** [comment]
I think perhaps this failure occurs if `cmake` is present, but you don't have a C++ compiler?

**Comment by xclaesse** [comment]
Could you run this with `export MESON_FORCE_BACKTRACE=1` to see where that error message comes from exactly?

**Comment by jon-turney** [comment]
```
2020-02-06T15:46:44.2731339Z Found pkg-config: /usr/bin/pkg-config (1.6.0)
2020-02-06T15:46:44.2731521Z Found CMake: /usr/bin/cmake (3.14.5)
2020-02-06T15:46:44.2731650Z 
2020-02-06T15:46:44.2731849Z testsuite/autotools/meson.build:8:0: ERROR: Unknown compiler(s): ['c++', 'g++', 'clang++', 'pgc++', 'icpc']
2020-02-06T15:46:44.2732092Z The follow exceptions were encountered:
2020-02-06T15:46:44.2732288Z Running "c++ --version" gave "[Errno 2] No such file or directory: 'c++': 'c++'"
2020-02-06T15:46:44.2732492Z Running "g++ --version" gave "[Errno 2] No such file or directory: 'g++': 'g++'"
2020-02-06T15:46:44.2732709Z Running "clang++ --version" gave "[Errno 2] No such file or directory: 'clang++': 'clang++'"
2020-02-06T15:46:44.2734189Z Running "pgc++ --version" gave "[Errno 2] No such file or directory: 'pgc++': 'pgc++'"
2020-02-06T15:46:44.2734519Z Running "icpc --version" gave "[Errno 2] No such file or directory: 'icpc': 'icpc'"
2020-02-06T15:46:44.2734663Z 
2020-02-06T15:46:44.2734874Z A full log can be found at /cygdrive/d/a/cygport/cygport/_build/meson-logs/meson-log.txt
2020-02-06T15:46:44.2948689Z Traceback (most recent call last):
2020-02-06T15:46:44.2949221Z   File "/usr/bin/meson", line 11, in <module>
2020-02-06T15:46:44.2949501Z     load_entry_point('meson==0.52.1', 'console_scripts', 'meson')()
2020-02-06T15:46:44.2949735Z   File "/usr/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 226, in main
2020-02-06T15:46:44.2949955Z     return run(sys.argv[1:], launcher)
2020-02-06T15:46:44.2950163Z   File "/usr/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 217, in run
2020-02-06T15:46:44.2950399Z     return CommandLineParser().run(args)
2020-02-06T15:46:44.2950603Z   File "/usr/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 129, in run
2020-02-06T15:46:44.2950813Z     return options.run_func(options)
2020-02-06T15:46:44.2951020Z   File "/usr/lib/python3.6/site-packages/mesonbuild/msetup.py", line 245, in run
2020-02-06T15:46:44.2951272Z     app.generate()
2020-02-06T15:46:44.2951483Z   File "/usr/lib/python3.6/site-packages/mesonbuild/msetup.py", line 159, in generate
2020-02-06T15:46:44.2951691Z     self._generate(env)
2020-02-06T15:46:44.2951897Z   File "/usr/lib/python3.6/site-packages/mesonbuild/msetup.py", line 192, in _generate
2020-02-06T15:46:44.2952103Z     intr.run()
2020-02-06T15:46:44.2952306Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 4024, in run
2020-02-06T15:46:44.2952509Z     super().run()
2020-02-06T15:46:44.2952700Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 412, in run
2020-02-06T15:46:44.2953111Z     self.evaluate_codeblock(self.ast, start=1)
2020-02-06T15:46:44.2953339Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 436, in evaluate_codeblock
2020-02-06T15:46:44.2953547Z     raise e
2020-02-06T15:46:44.2953759Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 430, in evaluate_codeblock
2020-02-06T15:46:44.2953976Z     self.evaluate_statement(cur)
2020-02-06T15:46:44.2954190Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 441, in evaluate_statement
2020-02-06T15:46:44.2954394Z     return self.function_call(cur)
2020-02-06T15:46:44.2954605Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 776, in function_call
2020-02-06T15:46:44.2954810Z     return func(node, posargs, kwargs)
2020-02-06T15:46:44.2955019Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 285, in wrapped
2020-02-06T15:46:44.2955220Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2955438Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 174, in wrapped
2020-02-06T15:46:44.2955654Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2955892Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 3542, in func_subdir
2020-02-06T15:46:44.2956105Z     self.evaluate_codeblock(codeblock)
2020-02-06T15:46:44.2956329Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 436, in evaluate_codeblock
2020-02-06T15:46:44.2956530Z     raise e
2020-02-06T15:46:44.2956731Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 430, in evaluate_codeblock
2020-02-06T15:46:44.2956931Z     self.evaluate_statement(cur)
2020-02-06T15:46:44.2957133Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 441, in evaluate_statement
2020-02-06T15:46:44.2957332Z     return self.function_call(cur)
2020-02-06T15:46:44.2957539Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 776, in function_call
2020-02-06T15:46:44.2958238Z     return func(node, posargs, kwargs)
2020-02-06T15:46:44.2958472Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 285, in wrapped
2020-02-06T15:46:44.2958674Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2958879Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 174, in wrapped
2020-02-06T15:46:44.2959077Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2959280Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 3542, in func_subdir
2020-02-06T15:46:44.2959478Z     self.evaluate_codeblock(codeblock)
2020-02-06T15:46:44.2959683Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 436, in evaluate_codeblock
2020-02-06T15:46:44.2959882Z     raise e
2020-02-06T15:46:44.2960072Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 430, in evaluate_codeblock
2020-02-06T15:46:44.2960273Z     self.evaluate_statement(cur)
2020-02-06T15:46:44.2960484Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 443, in evaluate_statement
2020-02-06T15:46:44.2960685Z     return self.assignment(cur)
2020-02-06T15:46:44.2960890Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 1054, in assignment
2020-02-06T15:46:44.2961092Z     value = self.evaluate_statement(node.value)
2020-02-06T15:46:44.2961300Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 441, in evaluate_statement
2020-02-06T15:46:44.2961497Z     return self.function_call(cur)
2020-02-06T15:46:44.2961701Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 776, in function_call
2020-02-06T15:46:44.2961905Z     return func(node, posargs, kwargs)
2020-02-06T15:46:44.2962092Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 285, in wrapped
2020-02-06T15:46:44.2962360Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2962568Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 285, in wrapped
2020-02-06T15:46:44.2962771Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2963090Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 285, in wrapped
2020-02-06T15:46:44.2963291Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2963482Z   [Previous line repeated 2 more times]
2020-02-06T15:46:44.2963682Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 155, in wrapped
2020-02-06T15:46:44.2963882Z     ret = f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2964091Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 174, in wrapped
2020-02-06T15:46:44.2964273Z     return f(*wrapped_args, **wrapped_kwargs)
2020-02-06T15:46:44.2964487Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 3092, in func_dependency
2020-02-06T15:46:44.2964695Z     d = self.dependency_impl(name, display_name, kwargs)
2020-02-06T15:46:44.2964924Z   File "/usr/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 3139, in dependency_impl
2020-02-06T15:46:44.2965142Z     dep = dependencies.find_external_dependency(name, self.environment, kwargs)
2020-02-06T15:46:44.2965372Z   File "/usr/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 2137, in find_external_dependency
2020-02-06T15:46:44.2965570Z     d = c()
2020-02-06T15:46:44.2965774Z   File "/usr/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 1061, in __init__
2020-02-06T15:46:44.2965985Z     CMakeDependency.class_cmakeinfo[self.for_machine] = self._get_cmake_info()
2020-02-06T15:46:44.2966207Z   File "/usr/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 1114, in _get_cmake_info
2020-02-06T15:46:44.2966422Z     ret1, out1, err1 = self._call_cmake(cmake_opts, 'CMakePathInfo.txt')
2020-02-06T15:46:44.2966902Z   File "/usr/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 1466, in _call_cmake
2020-02-06T15:46:44.2967140Z     return self.cmakebin.call_with_fake_build(args, build_dir, env=env)
2020-02-06T15:46:44.2967363Z   File "/usr/lib/python3.6/site-packages/mesonbuild/cmake/executor.py", line 198, in call_with_fake_build
2020-02-06T15:46:44.2967565Z     cxx_comp, cxx_launcher = choose_compiler('cpp')
2020-02-06T15:46:44.2967775Z   File "/usr/lib/python3.6/site-packages/mesonbuild/cmake/executor.py", line 185, in choose_compiler
2020-02-06T15:46:44.2967990Z     comp_obj = self.environment.compiler_from_language(lang, MachineChoice.BUILD)
2020-02-06T15:46:44.2968215Z   File "/usr/lib/python3.6/site-packages/mesonbuild/environment.py", line 1392, in compiler_from_language
2020-02-06T15:46:44.2968415Z     comp = self.detect_cpp_compiler(for_machine)
2020-02-06T15:46:44.2968626Z   File "/usr/lib/python3.6/site-packages/mesonbuild/environment.py", line 1027, in detect_cpp_compiler
2020-02-06T15:46:44.2968841Z     return self._detect_c_or_cpp_compiler('cpp', for_machine)
2020-02-06T15:46:44.2969061Z   File "/usr/lib/python3.6/site-packages/mesonbuild/environment.py", line 1021, in _detect_c_or_cpp_compiler
2020-02-06T15:46:44.2969270Z     self._handle_exceptions(popen_exceptions, compilers)
2020-02-06T15:46:44.2969465Z   File "/usr/lib/python3.6/site-packages/mesonbuild/environment.py", line 738, in _handle_exceptions
2020-02-06T15:46:44.2969669Z     raise EnvironmentException(errmsg)
2020-02-06T15:46:44.2969898Z mesonbuild.mesonlib.EnvironmentException: Unknown compiler(s): ['c++', 'g++', 'clang++', 'pgc++', 'icpc']
2020-02-06T15:46:44.2970099Z The follow exceptions were encountered:
2020-02-06T15:46:44.2970306Z Running "c++ --version" gave "[Errno 2] No such file or directory: 'c++': 'c++'"
2020-02-06T15:46:44.2970523Z Running "g++ --version" gave "[Errno 2] No such file or directory: 'g++': 'g++'"
2020-02-06T15:46:44.2970794Z Running "clang++ --version" gave "[Errno 2] No such file or directory: 'clang++': 'clang++'"
2020-02-06T15:46:44.2971016Z Running "pgc++ --version" gave "[Errno 2] No such file or directory: 'pgc++': 'pgc++'"
2020-02-06T15:46:44.2971232Z Running "icpc --version" gave "[Errno 2] No such file or directory: 'icpc': 'icpc'"
2020-02-06T15:46:44.5377254Z ##[error]Process completed with exit code 1.
```

**Comment by DimStar77** [comment]
> I think perhaps this failure occurs if `cmake` is present, but you don't have a C++ compiler?

That is true - but clearly cmake does not mandate c++; it's common, but not always true.

**Comment by jon-turney** [comment]
@mensinda ?

**Comment by mensinda** [comment]
Should be fixed with #6614

**Comment by DimStar77** [comment]
> Should be fixed with #6614

A first quick check succeeds:

```
Found pkg-config: /usr/bin/pkg-config (1.6.3)
Found CMake: /usr/bin/cmake (3.16.2)
Run-time dependency doesnotexist found: NO (tried pkgconfig and cmake)
Build targets in project: 0

Found ninja-1.9.0 at /usr/bin/ninja
```
